realm/src/main/java/io/realm/internal/Table.java

```diff
@@ -390,6 +390,54 @@ public long addEmptyRow() {
         return nativeAddEmptyRow(nativePtr, 1);
     }
 
+    public long addEmptyRowWithPrimaryKey(Object primaryKeyValue) {
+        checkImmutable();
+        if (hasPrimaryKey()) {
+            long primaryKeyColumnIndex = getPrimaryKey();
+            ColumnType type = getColumnType(primaryKeyColumnIndex);
+            long rowIndex;
+            Row row;
+
+            // Add with primary key initially set
+            switch(type) {
+                case STRING:
+                    if (!(primaryKeyValue instanceof String)) {
+                        throw new IllegalArgumentException("Primary key value is not a String: " + primaryKeyValue);
+                    }
+                    if (findFirstString(primaryKeyColumnIndex, (String)primaryKeyValue) != NO_MATCH) {
+                        throwDuplicatePrimaryKeyException(primaryKeyValue);
+                    }
+                    rowIndex = nativeAddEmptyRow(nativePtr, 1);
+                    row = getRow(rowIndex);
+                    row.setString(primaryKeyColumnIndex, (String) primaryKeyValue);
+                    break;
+
+                case INTEGER:
+                    long pkValue;
+                    try {
+                        pkValue = new Long(primaryKeyValue.toString());
+                    } catch (RuntimeException e) {
+                        throw new IllegalArgumentException("Primary key value is not a long: " + primaryKeyValue);
+                    }
+                    if (findFirstLong(primaryKeyColumnIndex, pkValue) != NO_MATCH) {
+                        throwDuplicatePrimaryKeyException(pkValue);
+                    }
+                    rowIndex = nativeAddEmptyRow(nativePtr, 1);
+                    row = getRow(rowIndex);
+                    row.setLong(primaryKeyColumnIndex, pkValue);
+                    break;
+
+                default:
+                    throw new RealmException("Cannot check for duplicate rows for unsupported primary key type: " + type);
+            }
+
+            return rowIndex;
+
+        } else {
```

Comment:
Minor style comment: I would encourage to check for the error conditions first. It gives less indenting in general. It would also be consistent with the style used in the case statements.